src/meili/map_matcher.cc

```diff
@@ -652,22 +653,33 @@ std::vector<MatchResults> MapMatcher::OfflineMatch(const std::vector<Measurement
   return best_paths;
 }
 
-std::unordered_map<StateId::Time, std::vector<Measurement>>
-MapMatcher::AppendMeasurements(const std::vector<Measurement>& measurements) {
+void MapMatcher::AppendMeasurements(
+    const std::vector<Measurement>& measurements,
+    std::unordered_map<StateId::Time, std::vector<Measurement>>& interpolated) {
   const float max_search_radius = config_.get<float>("max_search_radius"),
-              sq_max_search_radius = max_search_radius * max_search_radius;
-  const float interpolation_distance = config_.get<float>("interpolation_distance"),
-              sq_interpolation_distance = interpolation_distance * interpolation_distance;
-  std::unordered_map<StateId::Time, std::vector<Measurement>> interpolated;
-
-  // Always match the first measurement
-  auto last = measurements.cbegin();
-  auto time = AppendMeasurement(*last, sq_max_search_radius);
+              interpolation_distance = config_.get<float>("interpolation_distance"),
+              breakage_distance = config_.get<float>("breakage_distance"),
+              sq_max_search_radius = max_search_radius * max_search_radius,
+              sq_interpolation_distance = interpolation_distance * interpolation_distance,
+              sq_breakage_distance = breakage_distance * breakage_distance;
+
+  auto last = measurements.cbegin() - 1;
+  uint32_t time;
   double interpolated_epoch_time = -1;
   for (auto m = std::next(last); m != measurements.end(); ++m) {
-    const auto sq_distance = GreatCircleDistanceSquared(*last, *m);
-    // Always match the last measurement and if its far enough away
-    if (sq_interpolation_distance < sq_distance || std::next(m) == measurements.end()) {
+    if (interrupt_) {
+      (*interrupt_)();
+    }
+    std::vector<baldr::PathLocation> candidates;
+    const auto sq_distance = m == measurements.cbegin()
+                                 ? sq_interpolation_distance + 1 // Always match the first measurement
+                                 : GreatCircleDistanceSquared(*last, *m);
+    // Do not interpolate this point if:
+    // - it's far enough away from the last uninterpolated measurement,
+    // - it's closer to the last uninterpolated measurement than the max breakage distance,
+    // - and never interpolate the last measurement
+    if ((sq_distance >= sq_interpolation_distance && sq_distance < sq_breakage_distance) ||
```

Comment:
Noting for myself before the weekend hits: there's a bug here that causes the whole match to fail when the first measurement has no candidates. Basically because we start from `last=measurements-1`  on L666 above and the first measurement has empty candidates, we never move `last` up to the current measurement.